webserver/rate_limiter.py

```diff
@@ -0,0 +1,150 @@
+    # synchronized clocks between the workers and the redis server do not cause problems.
+    expiration_window = 10
+
+    def __init__(self, key_prefix, limit, per):
+        self.reset = (int(time.time()) // per) * per + per
+        self.key = key_prefix + str(self.reset)
+        self.limit = limit
+        self.per = per
+        p = _redis.pipeline()
+        p.incr(self.key)
+        p.expireat(self.key, self.reset + self.expiration_window)
+        self.current = p.execute()[0]
+
+    remaining = property(lambda x: max(x.limit - x.current, 0))
+    over_limit = property(lambda x: x.current > x.limit)
+
+def get_view_rate_limit():
+    return getattr(g, '_view_rate_limit', None)
+
+def on_over_limit(limit):
+    return 'You have exceeded your rate limit. See the X-RateLimit-* response headers for more ' \
+           'information on your current rate limit.\n', 429
+
+def set_rate_limits(per_ip, per_token, window):
+    _redis.put(RATELIMIT_PER_TOKEN_KEY, per_token)
+    _redis.put(RATELIMIT_PER_IP_KEY, per_ip)
+    _redis.put(RATELIMIT_WINDOW_KEY, window)
+
+def check_limit_freshness():
+    limits_timeout = getattr(g, '_' + RATELIMIT_TIMEOUT, 0)
+    if time.time() <= limits_timeout:
+        return
+
+    value = int(_redis.get(RATELIMIT_PER_TOKEN_KEY) or '0')
+    if not value:
+        _redis.set(RATELIMIT_PER_TOKEN_KEY, RATELIMIT_PER_TOKEN_DEFAULT)
+        value = RATELIMIT_PER_TOKEN_DEFAULT
+    setattr(g, '_' + RATELIMIT_PER_TOKEN_KEY, value)
+
+    value = int(_redis.get(RATELIMIT_PER_IP_KEY) or '0')
+    if not value:
+        _redis.set(RATELIMIT_PER_IP_KEY, RATELIMIT_PER_IP_DEFAULT)
+        value = RATELIMIT_PER_IP_DEFAULT
+    setattr(g, '_' + RATELIMIT_PER_IP_KEY, value)
+
+    value = int(_redis.get(RATELIMIT_WINDOW_KEY) or '0')
+    if not value:
+        _redis.set(RATELIMIT_WINDOW_KEY, RATELIMIT_WINDOW_DEFAULT)
+        value = RATELIMIT_WINDOW_DEFAULT
+    setattr(g, '_' + RATELIMIT_WINDOW_KEY, value)
+
+    setattr(g, '_' + RATELIMIT_TIMEOUT, int(time.time()) + RATELIMIT_REFRESH)
+
+def get_per_ip_limits():
+    check_limit_freshness()
+    return {
+            'limit':   getattr(g, '_' + RATELIMIT_PER_IP_KEY), 
+            'window' : getattr(g, '_' + RATELIMIT_WINDOW_KEY),
+           }
+
+def get_per_token_limits():
+    check_limit_freshness()
+    return {
+            'limit':   getattr(g, '_' + RATELIMIT_PER_TOKEN_KEY), 
+            'window' : getattr(g, '_' + RATELIMIT_WINDOW_KEY),
+           }
+
+def get_rate_limit_data(request):
+    '''Fetch key for the given request. If an Authorization header is provided, 
+       the caller will get a better and personalized rate limit. If no header is provided,
+       the caller will be rate limited by IP, which gets an overall lower rate limit.
+       This should encourage callers to always provide the Authorization token
+    '''
+
+    auth_token = request.headers.get('Authorization')
+    if auth_token:
+        auth_token = auth_token[6:]
```

Comment:
I'd say `auth_header` would be a less confusing variable name to use above. And then `auth_token` on this line and further when it's the actual token inside.